cpp/include/cudf/io/datasource.hpp

```diff
@@ -36,6 +37,242 @@ namespace io {
+   * multi-threading or asynchronous operations are supported.
+   *
+   * The primary purpose of this datasource type is to be a base class for the
+   * `O_DIRECT` implementation, which needs to issue pread() calls against a
+   * file descriptor that *hasn't* been opened with `O_DIRECT` if certain
+   * constraints aren't met (specifically: when reading the final bytes of a
+   * file that isn't perfectly aligned to a sector-size boundary).
+   *
+   * The time required to service reads from this data source will be affected
+   * by the presence or absence of the desired data in the Linux page cache.
+   * Thus, back-to-back runs of the same file will have significantly different
+   * performance characteristics, depending on whether the data is in the page
+   * cache or not.
+   *
+   * Generally, this data source should be avoided in favor of the `KVIKIO`
+   * data source, which will be more performant in most cases.  Thus, it can
+   * be used as a baseline for which improved `KVIKIO` performance can be
+   * empirically measured.
+   */
+  HOST,
+
+  /**
+   * @brief Host-based data source that issues reads against a file descriptor
+   * opened with `O_DIRECT`, where possible, bypassing the Linux page cache.
+   *
+   * This data source will always result in the slowest possible read times,
+   * as all reads are serviced directly from the underlying device.  However,
+   * it will be consistently slow, and that consistency can be critical when
+   * benchmarking or profiling changes purporting to improve performance in
+   * unrelated areas.
+   *
+   * Thus, the primary use case for this data source is for benchmarking and
+   * profiling purposes, where you want to eliminate any runtime variance in
+   * back-to-back runs that would be caused by the presence or absence of data
+   * in the host's page cache.
+   *
+   * A secondary use case for this data source is when you specifically do not
+   * want to pollute the host's page cache with the data being read, either
+   * because it won't be read again soon, or you want to remove the memory
+   * pressure (or small but non-trivial amount of compute overhead) that would
+   * otherwise be introduced by servicing I/O through the page cache.  In some
+   * scenarios, this can yield a net performance improvement, despite a higher
+   * per-read latency.
+   *
+   * A real-life example of how this can manifest is when doing very large TPC-H
+   * or TPC-DS runs, where the data set is orders of magnitude larger than the
+   * available host memory, e.g. 30TB or 100TB runs on hosts with <= 4TB of RAM.
+   *
+   * For certain queries--typically read-heavy, join-heavy ones--doing `O_DIRECT`
+   * reads can result in a net performance improvement, as the host's page cache
+   * won't be polluted with data that will never be read again, and compute
+   * overhead associated with cache thrashing when memory is tight is
+   * eliminated.
+   */
+  ODIRECT,
```

Comment:
Would you mind splitting the O_DIRECT source into a separate PR? I think it's fair to say that it's a separate feature (dependent of the `create` API change).

Reply:
Actually, these are more intertwined than I realized. Need to review this very carefully :)

Reply:
It would definitely be possible to hoist-out O_DIRECT into a separate PR.  Or possibly separate commits to ease review.  But yeah things are pretty interrelated.